File: frontend/src/metabase/api/wso2.ts
import type { Settings } from "metabase-types/api";

import { Api } from "./api";
import { invalidateTags, tag } from "./tags";

type Wso2AuthSettings = Pick<
  Settings,
  | "wso2-enabled"
  | "wso2-issuer-uri"
  | "wso2-client-id"
  | "wso2-client-secret"
  | "wso2-attribute-email"
  | "wso2-attribute-firstname"
  | "wso2-attribute-lastname"
>;

export const wso2Api = Api.injectEndpoints({
  endpoints: (builder) => ({
    updateWso2Auth: builder.mutation<void, Wso2AuthSettings>({
      query: (settings) => ({
        method: "PUT",
        url: `/api/wso2/settings`,
        body: settings,
      }),
      invalidatesTags: (_, error) =>
        invalidateTags(error, [tag("session-properties")]),
    }),
  }),
});

export const { useUpdateWso2AuthMutation } = wso2Api;

File: src/metabase/api/wso2.clj
(ns metabase.api.wso2
  "/api/wso2 endpoints"
  (:require
   [metabase.api.common :as api]
   [metabase.api.macros :as api.macros]
   [metabase.settings.core :as setting]
   [metabase.sso.settings :as sso-settings]))

#_{:clj-kondo/ignore [:metabase/validate-defendpoint-has-response-schema]}
(api.macros/defendpoint :put "/settings"
  "Update WSO2 Identity Server SSO settings. You must be a superuser to do this."
  [_route-params
   _query-params
   {:keys [wso2-client-id wso2-client-secret wso2-issuer-uri
           wso2-attribute-email wso2-attribute-firstname wso2-attribute-lastname
           wso2-enabled]
    :as body}
   :- [:map
       [:wso2-client-id {:optional true} [:maybe :string]]
       [:wso2-client-secret {:optional true} [:maybe :string]]
       [:wso2-issuer-uri {:optional true} [:maybe :string]]
       [:wso2-attribute-email {:optional true} [:maybe :string]]
       [:wso2-attribute-firstname {:optional true} [:maybe :string]]
       [:wso2-attribute-lastname {:optional true} [:maybe :string]]
       [:wso2-enabled {:optional true} [:maybe :boolean]]]]
  (api/check-superuser)
  (setting/set-many! body)
  (when (contains? body :wso2-enabled)
    (sso-settings/wso2-enabled! (:wso2-enabled body))))

(def ^{:arglists '([request respond raise])} routes
  "`/api/wso2/` routes."
  (api.macros/ns-handler *ns*))
